File: kube/src/main/java/com/jacksonrakena/infrastructure/apps/Jacksonbot.kt
package com.jacksonrakena.infrastructure.apps

import com.jacksonrakena.infrastructure.util.applyCommonConfiguration
import org.cdk8s.Chart
import org.cdk8s.ChartProps
import org.cdk8s.plus28.ContainerPort
import org.cdk8s.plus28.ContainerProps
import org.cdk8s.plus28.Deployment
import org.cdk8s.plus28.DeploymentProps
import org.cdk8s.plus28.DeploymentStrategy
import org.cdk8s.plus28.DockerConfigSecret
import org.cdk8s.plus28.EnvValue
import org.cdk8s.plus28.IConfigMap
import org.cdk8s.plus28.Service
import org.cdk8s.plus28.Volume
import org.cdk8s.plus28.VolumeMount
import software.constructs.Construct

class Jacksonbot(
    scope: Construct,
    id: String,
    configMap: IConfigMap,
    registrySecret: DockerConfigSecret,
    postgresService: Service,
    props: ChartProps? = null,
) : Chart(scope, id, props) {
    val volume = Volume.fromConfigMap(
        this,
        "jacksonbot-config-mount",
        configMap
    )
    val deployment = Deployment(
        this,
        "deployment",
        DeploymentProps.builder()
            .replicas(1)
            .strategy(
                DeploymentStrategy.recreate()
            )
            .volumes(
                listOf(
                    volume
                )
            )
            .containers(
                listOf(
                    ContainerProps.builder()
                        .name("bot")
                        .image("ghcr.io/jacksonrakena/jacksonbot:latest")
                        .applyCommonConfiguration()
                        .ports(
                            listOf(
                                ContainerPort.builder().number(80).build()
                            )
                        )
                        .volumeMounts(
                            listOf(
                                VolumeMount.builder().volume(volume).path("/app/jacksonbot.appsettings.json")
                                    .subPath("jacksonbot.appsettings.json").build()
                            )
                        )
                        .envVariables(
                            mapOf(
                                "JACKSONBOT_ConnectionStrings__Database" to EnvValue.fromValue("Host=${postgresService.name};UserName=jacksonbot;Password=jacksonbot;Database=jacksonbot")
                            )
                        )
                        .build()
                )
            )
            .dockerRegistryAuth(registrySecret)
            .build()
    )
}

File: src/apps/jacksonbot.ts
import { Construct } from "constructs";
import { Chart, ChartProps } from "cdk8s";
import * as kplus from "cdk8s-plus-28";
import { applyCommonConfiguration } from "../util/container-utils";

export class Jacksonbot extends Chart {
  constructor(
    scope: Construct,
    id: string,
    configMap: kplus.IConfigMap,
    registrySecret: kplus.DockerConfigSecret,
    postgresService: kplus.Service,
    props?: ChartProps
  ) {
    super(scope, id, props);

    const volume = kplus.Volume.fromConfigMap(
      this,
      "jacksonbot-config-mount",
      configMap
    );

    new kplus.Deployment(this, "deployment", {
      replicas: 1,
      strategy: kplus.DeploymentStrategy.recreate(),
      volumes: [volume],
      containers: [
        applyCommonConfiguration({
          name: "bot",
          image: "ghcr.io/jacksonrakena/jacksonbot:latest",
          ports: [{ number: 80 }],
          volumeMounts: [
            {
              volume,
              path: "/app/jacksonbot.appsettings.json",
              subPath: "jacksonbot.appsettings.json",
            },
          ],
          envVariables: {
            JACKSONBOT_ConnectionStrings__Database: kplus.EnvValue.fromValue(
              `Host=${postgresService.name};UserName=jacksonbot;Password=jacksonbot;Database=jacksonbot`
            ),
          },
        }),
      ],
      dockerRegistryAuth: registrySecret,
    });
  }
}
